pkg/retry/retry.go

```diff
@@ -24,6 +26,42 @@ type Ticker interface {
 	Stop()
 }
 
+// ErrorSet represents a set of unique errors.
+type ErrorSet struct {
+	errs map[string]error
+
+	mu sync.Mutex
+}
+
+func (e *ErrorSet) Error() string {
+	if len(e.errs) == 0 {
+		return ""
+	}
+
+	errString := fmt.Sprintf("%d error(s) occurred:", len(e.errs))
+	for _, err := range e.errs {
+		errString = fmt.Sprintf("%s\n%s", errString, err)
+	}
+
+	return errString
+}
+
+// Append adds the error to the set if the error is not already present.
+func (e *ErrorSet) Append(err error) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if e.errs == nil {
+		e.errs = make(map[string]error)
+	}
+
+	if _, ok := e.errs[err.Error()]; !ok {
+		e.errs[err.Error()] = err
+	}
+
+	return e
+}
```

Comment:
I wonder if we should use a slice to preserve order.

Reply:
I'm not sure if order matters after deduplication

Reply:
SGTM